src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/dcoff.cpp

```diff
@@ -21,6 +21,97 @@
 namespace ov {
 namespace npuw {
 
+namespace pattern_utils {
+
+std::shared_ptr<ov::op::v0::MatMul> find_matmul_downwards(const std::shared_ptr<ov::Node>& start_node) {
+    std::shared_ptr<ov::Node> current_node = start_node;
+    while (current_node) {
+        // Check if the current node is a MatMul
+        if (auto matmul = std::dynamic_pointer_cast<ov::op::v0::MatMul>(current_node)) {
+            return matmul;
+        }
+        // Move to the next node in the path if there is one
+        if (!current_node->outputs().empty()) {
+            auto output = current_node->outputs().at(0);
+            if (!output.get_target_inputs().empty()) {
+                current_node = output.get_target_inputs().begin()->get_node()->shared_from_this();
+            } else {
+                // No further outputs, end the search
+                break;
+            }
+        } else {
+            // No outputs, end the search
+            break;
+        }
+    }
+    return nullptr; // MatMul not found
+}
+
+
+std::shared_ptr<ov::op::v0::MatMul> get_root_matmul(ov::pass::pattern::Matcher& m) {
+    auto& node_to_output = m.get_pattern_value_map();
+
+    // If the map is not empty, start the search from the first matched node
+    if (!node_to_output.empty()) {
+        auto start_node = node_to_output.begin()->second.get_node_shared_ptr();
+        return find_matmul_downwards(start_node);
+    }
+
+    // If the map is empty or no MatMul node is found, return nullptr
+    LOG_DEBUG("NO MATMUL FOUND!");
+    return nullptr;
+}
+
+bool transpose_required(const std::shared_ptr<ov::op::v0::MatMul>& matmul_node) {
+    if (!matmul_node) {
+        LOG_DEBUG("NOT a MATMUL NODE!");
+    }
+
+    // Get the shape of the second input to the MatMul node
+    const auto& input_shape = matmul_node->input_value(1).get_shape();
+
+    // Check if the highest dimension is not at the first index
+    if (input_shape.size() >= 2) {
+        size_t max_dim = *std::max_element(input_shape.begin(), input_shape.end());
+        if (input_shape[0] != max_dim) {
+            return true; // Transpose is required
+        }
+    }
+
+    return false; // Transpose is not required
+}
+
+void transpose_param_shape(std::shared_ptr<ov::op::v0::Parameter>& param) {
+    auto partial_shape = param->get_partial_shape();
+
+    // Ensure the shape is static before proceeding
+    if (partial_shape.is_static()) {
+        auto shape = partial_shape.to_shape();
+
+        // Check if the shape is 2D or 3D and needs transposing
+        if (shape.size() == 2 && shape[0] < shape[1]) {
```

Comment:
I believe we shouldn't look at `<` or max_elements here. The requirement is not to place the max dim to the proper dimension. Please check the issue description

Reply:
(Stopped review at this point)

Reply:
Rectified!